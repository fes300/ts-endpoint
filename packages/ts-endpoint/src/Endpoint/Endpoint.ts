import * as t from 'io-ts';
import { identity } from 'fp-ts/lib/function';
import { RequiredKeys } from 'typelevel-ts';

export type PropsType<P> = P extends {} ? { [k in RequiredKeys<P>]: t.TypeOf<P[k]> } : never;

export type EndpointInput<E extends EndpointInstance<any>> = PropsType<E['Input']>;
export type EndpointOutput<E extends EndpointInstance<any>> = t.TypeOf<E['Output']>;

export const HTTPMethod = t.keyof(
  {
    OPTIONS: null,
    HEAD: null,
    GET: null,
    POST: null,
    PUT: null,
    PATCH: null,
    DELETE: null,
  },
  'HTTPMethod'
);
export type HTTPMethod = t.TypeOf<typeof HTTPMethod>;

/**
 * Endpoint options type
 */
export interface EndpointOptions {
  stringifyBody?: boolean;
}

export const defaultOps = {
  stringifyBody: false,
};

/**
 * Represents an endpoint of our API
 */
export interface Endpoint<
  M extends HTTPMethod,
  O extends t.Type<any, any, any>,
  H extends { [k: string]: t.Type<any, any, any> } | undefined = undefined,
  Q extends { [k: string]: t.Type<any, any, any> } | undefined = undefined,
  B extends t.Type<any, any, any> | undefined = undefined,
  P extends { [k: string]: t.Type<any, any, any> } | undefined = undefined
> {
  /* utils to get the full path given a set of query params */
  getPath: P extends undefined
    ? () => string
    : (args: { [k in keyof P]: P[k] extends t.Any ? t.TypeOf<P[k]> : never }) => string;
  Method: M;
  Opts?: EndpointOptions;
  Input: {
    Headers?: H;
    Params?: P;
    Query?: Q;
    Body?: M extends 'POST' | 'PUT' | 'PATCH' ? B : never;
  };
  Output: O;
}

/**
 * Data type representing an endpoint instance.
 * @public getStaticPath accepts a formatting function (param: string) -> string and returns
 **/
export type EndpointInstance<E extends Endpoint<any, any, any, any, any, any>> = {
  /**
   * helper to get a path given a set of runtime params.
   *
   * @returns a string representation of a path instance
   * @example
   * ```
   * import { Endpoint } from '..';
   * import * as t from 'io-ts';
   *
   * const endpoint = Endpoint({
   *  Input: {
   *    Params: { id: t.number },
   *  },
   *  Method: 'GET',
   *  getPath: ({ id }) => `users/${id}/crayons`,
   *  Output: { crayons: t.array(t.string) },
   *});
   *
   * endpoint.getPath({ id: 3 })
   * // returns "users/3/crayons"
   * ```
   */
  getPath: E['getPath'];
  /**
   * helper to get a path version with static values in place of actual params.
   *
   * @returns a static representation of the path
   * @example
   * ```
   * import { Endpoint } from '..';
   * import * as t from 'io-ts';
   *
   * const endpoint = Endpoint({
   *  Input: {
   *    Params: { id: t.string },
   *  },
   *  Method: 'GET',
   *  getPath: ({ id }) => `users/${id}/crayons`,
   *  Output: { crayons: t.array(t.string) },
   *});
   *
   * endpoint.getStaticPath(param => `:${param}`) // returns "users/:id/crayons"
   * ```
   */
  getStaticPath: E['Input']['Params'] extends undefined
    ? () => string
    : (f: (paramName: string) => string) => string;
  Method: E['Method'];
  Output: E['Output'];
  Opts: EndpointOptions;
  Input: (E['Input']['Params'] extends undefined
    ? { Params?: never }
    : { Params: t.TypeC<NonNullable<E['Input']['Params']>> }) &
    (E['Input']['Headers'] extends undefined
      ? { Headers?: never }
      : { Headers: t.TypeC<NonNullable<E['Input']['Headers']>> }) &
    (E['Input']['Query'] extends undefined
      ? { Query?: never }
      : { Query: t.TypeC<NonNullable<E['Input']['Query']>> }) &
    (E['Input']['Body'] extends undefined
      ? { Body?: never }
      : { Body: NonNullable<E['Input']['Body']> });
};

export type TypeOfEndpointInstance<E extends EndpointInstance<any>> = {
  getPath: E['getPath'];
  getStaticPath: E['getStaticPath'];
  Opts: E['Opts'];
  Method: E['Method'];
  Output: t.TypeOf<E['Output']>;
  Input: {
    [k in keyof E['Input']]: E['Input'][k] extends t.Type<any, any, any>
      ? t.TypeOf<E['Input'][k]>
      : never;
  };
};

const addSlash = (s: string) => (s.substr(0, 1) === '/' ? s : `/${s}`);

/**
 * Constructor function for an endpoint
 * @returns an EndpointInstance
 */
export function Endpoint<
  M extends HTTPMethod,
  O extends t.Type<any, any, any>,
  H extends { [k: string]: t.Type<any, any, any> } | undefined = undefined,
  Q extends { [k: string]: t.Type<any, any, any> } | undefined = undefined,
  B extends t.Type<any, any, any> | undefined = undefined,
  P extends { [k: string]: t.Type<any, any, any> } | undefined = undefined
>(e: Endpoint<M, O, H, Q, B, P>): EndpointInstance<Endpoint<M, O, H, Q, B, P>> {
  return ({
    ...e,
    getPath: ((i) => {
      const path = e.getPath(i);
      return addSlash(path);
    }) as typeof e.getPath,
    Output: e.Output,
    getStaticPath: (f: P extends undefined ? undefined : (paramName: string) => string) => {
      const path = e.getPath(
        Object.keys(e.Input.Params ?? {}).reduce(
          (acc, k) => ({ ...acc, [k]: (f ?? identity)(k) }),
          {}
        )
      );
      return addSlash(path);
    },
    Input: {
      ...(e.Input.Body !== undefined ? { Body: e.Input.Body } : {}),
      ...(e.Input.Headers !== undefined ? { Headers: t.type(e.Input.Headers as t.Props, 'Headers') } : {}),
      ...(e.Input.Params !== undefined ? { Params: t.type(e.Input.Params as t.Props, 'Params') } : {}),
      ...(e.Input.Query !== undefined ? { Query: t.type(e.Input.Query as t.Props, 'Query') } : {}),
    },
    Opts: e.Opts ?? defaultOps,
  } as unknown) as EndpointInstance<Endpoint<M, O, H, Q, B, P>>;
}
