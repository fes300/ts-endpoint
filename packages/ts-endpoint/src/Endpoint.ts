import { pipe } from 'fp-ts/function';
import * as R from 'fp-ts/Record';
import { Codec, RecordCodec, runtimeType } from 'ts-io-error/lib/Codec';
import { MinimalEndpoint } from '.';
import { addSlash, InferEndpointParams } from './helpers';

export type HTTPMethod = 'OPTIONS' | 'HEAD' | 'GET' | 'POST' | 'PUT' | 'PATCH' | 'DELETE';

/**
 * Represents an HTTP endpoint of our API
 */
export interface Endpoint<
  M extends HTTPMethod,
  O extends Codec<any, any, any>,
  H extends RecordCodec<any, any> | undefined = undefined,
  Q extends RecordCodec<any, any> | undefined = undefined,
  B extends Codec<any, any, any> | undefined = undefined,
  P extends RecordCodec<any, any> | undefined = undefined,
  E extends EndpointErrors<never, Codec<any, any, any>> | undefined = undefined
> {
  /* utils to get the full path given a set of query params */
  getPath: [P] extends [undefined] ? (i?: {}) => string : (args: runtimeType<P>) => string;
  Method: M;
  Errors?: E;
  Input?: [H, P, Q, B] extends [undefined, undefined, undefined, undefined]
    ? never
    : {
        Headers?: H;
        Params?: P;
        Query?: Q;
        Body?: M extends 'POST' | 'PUT' | 'PATCH' | 'DELETE' ? B : never;
      };
  Output: O;
}

/**
 * Data type representing an endpoint instance.
 **/
export type EndpointInstance<E extends MinimalEndpoint> = {
  /**
   * helper to get a path given a set of runtime params.
   *
   * @returns a string representation of a path instance
   * @example
   * ```
   * import { Endpoint } from '..';
   * import * as t from 'io-ts';
   *
   * const endpoint = Endpoint({
   *  Input: {
   *    Params: { id: t.number },
   *  },
   *  Method: 'GET',
   *  getPath: ({ id }) => `users/${id}/crayons`,
   *  Output: { crayons: t.array(t.string) },
   *});
   *
   * endpoint.getPath({ id: 3 })
   * // returns "users/3/crayons"
   * ```
   */
  getPath: E['getPath'];
  /**
   * helper to get a path version with static values in place of actual params.
   *
   * @returns a static representation of the path
   * @example
   * ```
   * import { Endpoint } from '..';
   * import * as t from 'io-ts';
   *
   * const endpoint = Endpoint({
   *  Input: {
   *    Params: { id: t.string },
   *  },
   *  Method: 'GET',
   *  getPath: ({ id }) => `users/${id}/crayons`,
   *  Output: { crayons: t.array(t.string) },
   *});
   *
   * endpoint.getStaticPath(param => `:${param}`) // returns "users/:id/crayons"
   * ```
   */
  getStaticPath: [E['Input']] extends [undefined]
    ? (i?: {}) => string
    : [InferEndpointParams<E>['params']] extends [undefined]
    ? (i?: {}) => string
    : (f: (paramName: keyof runtimeType<InferEndpointParams<E>['params']>) => string) => string;
  Method: E['Method'];
  Output: E['Output'];
} & (E['Input'] extends undefined
  ? {
      Input?: never;
    }
  : {
      Input: (InferEndpointParams<E>['params'] extends undefined
        ? { Params?: never }
        : { Params: NonNullable<InferEndpointParams<E>['params']> }) &
        (InferEndpointParams<E>['headers'] extends undefined
          ? { Headers?: never }
          : { Headers: NonNullable<InferEndpointParams<E>['headers']> }) &
        (InferEndpointParams<E>['query'] extends undefined
          ? { Query?: never }
          : { Query: NonNullable<InferEndpointParams<E>['query']> }) &
        (InferEndpointParams<E>['body'] extends undefined
          ? { Body?: never }
          : { Body: NonNullable<InferEndpointParams<E>['body']> });
    }) &
  (E['Errors'] extends undefined ? { Errors?: never } : { Errors: E['Errors'] });

export type EndpointErrors<S extends string, B extends Codec<any, any, any>> = Record<S, B>;

/**
 * Constructor function for an endpoint
 * @returns an EndpointInstance
 */
export function Endpoint<
  M extends HTTPMethod,
  O extends Codec<any, any, any>,
  H extends RecordCodec<any, any> | undefined = undefined,
  Q extends RecordCodec<any, any> | undefined = undefined,
  B extends Codec<any, any, any> | undefined = undefined,
  P extends RecordCodec<any, any> | undefined = undefined,
  E extends EndpointErrors<never, Codec<any, any, any>> | undefined = undefined
>(e: Endpoint<M, O, H, Q, B, P, E>): EndpointInstance<Endpoint<M, O, H, Q, B, P, E>> {
  // TODO: check if the headers are valid?
  // const headersWithWhiteSpaces = pipe(
  //   e.Input?.Headers?.props ?? {},
  //   R.filterWithIndex((k: string) => k.indexOf(' ') !== -1),
  //   R.keys
  // );

  // if (headersWithWhiteSpaces.length > 0) {
  //   console.error('white spaces are not allowed in Headers names:', headersWithWhiteSpaces);
  // }

  return ({
    ...e,
    getPath: ((i: any) => {
      const path = e.getPath(i);

      return addSlash(path);
    }) as typeof e.getPath,
    getStaticPath: (f: (paramName: any) => string) => {
      const params = e.Input?.Params;

      if (params === undefined) {
        return addSlash(e.getPath(undefined as any));
      }

      return pipe(
        params.props,
        R.mapWithIndex((k) => (f ? f(k) : k)),
        (v: any) => e.getPath(v),
        addSlash
      );
    },
    Output: e.Output,
    ...(e.Errors ? { Errors: e.Errors } : {}),
    Input: {
      ...(e.Input?.Body ? { Body: e.Input.Body } : {}),
      ...(e.Input?.Headers ? { Headers: e.Input.Headers } : {}),
      ...(e.Input?.Params ? { Params: e.Input.Params } : {}),
      ...(e.Input?.Query ? { Query: e.Input.Query } : {}),
    },
  } as unknown) as EndpointInstance<Endpoint<M, O, H, Q, B, P, E>>;
}
