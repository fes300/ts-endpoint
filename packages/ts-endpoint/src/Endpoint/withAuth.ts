import * as t from 'io-ts';
import { BearerToken } from 'ts-shared/lib/BearerToken';
import { EndpointInstance } from '.';
import { Endpoint } from './Endpoint';

export const AuthHeader = {
  authorization: BearerToken,
};

type DefaultEmpty<A> = A extends undefined ? {} : A;

type WithAuth<E> = E extends EndpointInstance<
  Endpoint<infer M, infer O, infer H, infer Q, infer B, infer P>
>
  ? EndpointInstance<Endpoint<M, O, DefaultEmpty<H> & typeof AuthHeader, Q, B, P>>
  : never;

export function withAuth<E extends EndpointInstance<any>>(e: E): WithAuth<E> {
  const originalHeadersProps: t.Props = e.Input.Headers?.props ?? {};
  const newHeaders = { ...originalHeadersProps, ...AuthHeader };

  return ({
    ...e,
    Input: {
      ...e.Input,
      Headers: t.type(newHeaders, 'Headers'),
    },
  } as unknown) as WithAuth<E>;
}
