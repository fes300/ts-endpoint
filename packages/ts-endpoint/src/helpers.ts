import { EndpointInstance, HTTPMethod } from '.';
import { RequiredKeys } from 'typelevel-ts';
import { Endpoint, EndpointErrors } from './Endpoint';
import { Codec, runtimeType, RecordCodec } from 'ts-io-error/lib/Codec';

export const addSlash = (s: string) => (s.substr(0, 1) === '/' ? s : `/${s}`);

export type MinimalEndpoint = Omit<
  Endpoint<
    HTTPMethod,
    Codec<any, any, any>,
    RecordCodec<any, any>,
    RecordCodec<any, any>,
    Codec<any, any, any>,
    RecordCodec<any, any>,
    EndpointErrors<never, Codec<any, any, any>>
  >,
  'getPath'
> & { getPath: (i?: any) => string };

export type MinimalEndpointInstance = MinimalEndpoint & {
  getStaticPath: (f: (i?: any) => string) => string;
};

export type TypeOfEndpointInstance<E extends MinimalEndpointInstance> = {
  getPath: E['getPath'];
  getStaticPath: E['getStaticPath'];
  Method: E['Method'];
  Output: runtimeType<E['Output']>;
  Errors: {
    [k in keyof NonNullable<E['Errors']>]: NonNullable<E['Errors']>[k] extends Codec<any, any, any>
      ? runtimeType<NonNullable<E['Errors']>[k]>
      : never;
  };
  Input: [RequiredKeys<E['Input']>] extends [never]
    ? void
    : {
        [k in keyof NonNullable<E['Input']>]: NonNullable<E['Input']>[k] extends Codec<
          any,
          any,
          any
        >
          ? runtimeType<NonNullable<E['Input']>[k]>
          : never;
      };
};

export type DecodedPropsType<P> = P extends {}
  ? { [k in RequiredKeys<P>]: runtimeType<P[k]> }
  : never;

export type KnownErrorStatus<W> = undefined extends W
  ? undefined
  : W extends Record<infer K, any>
  ? K
  : undefined;

export type KnownErrorBody<W> = undefined extends W
  ? undefined
  : W extends Record<any, infer V>
  ? V
  : undefined;

export type InferEndpointParams<E> = E extends Endpoint<
  infer M,
  infer O,
  infer H,
  infer Q,
  infer B,
  infer P,
  infer E
>
  ? { method: M; output: O; headers: H; query: Q; body: B; params: P; errors: E }
  : never;

export type InferEndpointInstanceParams<EI> = EI extends EndpointInstance<infer E>
  ? InferEndpointParams<E>
  : never;

export declare type UndefinedOrRuntime<N> = N extends RecordCodec<any, any>
  ? {
      [k in keyof N['props']]: runtimeType<N['props'][k]>;
    }
  : undefined;