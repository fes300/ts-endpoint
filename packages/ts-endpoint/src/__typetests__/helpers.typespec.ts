import * as t from 'io-ts';
import { RequiredKeys } from 'typelevel-ts';
import { Endpoint } from '../Endpoint';
import { TypeOfEndpointInstance } from '../helpers';

const endpointWithParam = Endpoint({
  Input: {
    Query: t.type({ color: t.string }),
    Params: t.type({ id: t.number }),
  },
  Method: 'GET',
  getPath: ({ id }) => `users/${id.toString()}/crayons`,
  Output: t.type({ crayons: t.array(t.string) }),
});

const endpointWithoutParam = Endpoint({
  Input: {
    Query: t.type({ color: t.string }),
  },
  Method: 'GET',
  getPath: () => `users/crayons`,
  Output: t.type({ crayons: t.array(t.string) }),
});

const endpointWithErrors = Endpoint({
  Input: {
    Query: t.type({ color: t.string }),
    Params: t.type({ id: t.number }),
  },
  Method: 'GET',
  getPath: ({ id }) => `users/${id.toString()}/crayons`,
  Output: t.type({ crayons: t.array(t.string) }),
  Errors: {
    401: t.undefined,
    404: t.type({ message: t.string }),
    500: t.type({ foo: t.number }),
  },
});

const endpointWithoutInput = Endpoint({
  Method: 'GET',
  getPath: () => `users/crayons`,
  Output: t.type({ crayons: t.array(t.string) }),
});

const endpointWithBody = Endpoint({
  Method: 'POST',
  getPath: () => `users/crayons`,
  Output: t.type({ crayons: t.array(t.string) }),
  Input: { Body: t.type({ id: t.number }) },
});

// @dts-jest:pass:snap resulting TypeOfEndpointInstance typings are correct
null as TypeOfEndpointInstance<typeof endpointWithParam>['Errors'];
// @dts-jest:pass:snap resulting TypeOfEndpointInstance typings are correct
null as TypeOfEndpointInstance<typeof endpointWithParam>['Input'];
// @dts-jest:pass:snap resulting TypeOfEndpointInstance typings are correct
null as TypeOfEndpointInstance<typeof endpointWithParam>['Output'];
// @dts-jest:pass:snap resulting TypeOfEndpointInstance typings are correct
null as TypeOfEndpointInstance<typeof endpointWithParam>['getPath'];
// @dts-jest:pass:snap resulting TypeOfEndpointInstance typings are correct
null as TypeOfEndpointInstance<typeof endpointWithParam>['getStaticPath'];
// @dts-jest:pass:snap resulting TypeOfEndpointInstance typings are correct
null as RequiredKeys<TypeOfEndpointInstance<typeof endpointWithParam>>;
// @dts-jest:pass:snap resulting TypeOfEndpointInstance typings are correct
null as RequiredKeys<TypeOfEndpointInstance<typeof endpointWithParam>['Input']>;

// @dts-jest:pass:snap resulting TypeOfEndpointInstance typings are correct
null as TypeOfEndpointInstance<typeof endpointWithoutParam>['Errors'];
// @dts-jest:pass:snap resulting TypeOfEndpointInstance typings are correct
null as TypeOfEndpointInstance<typeof endpointWithoutParam>['Input'];
// @dts-jest:pass:snap resulting TypeOfEndpointInstance typings are correct
null as TypeOfEndpointInstance<typeof endpointWithoutParam>['Output'];
// @dts-jest:pass:snap resulting TypeOfEndpointInstance typings are correct
null as TypeOfEndpointInstance<typeof endpointWithoutParam>['getPath'];
// @dts-jest:pass:snap resulting TypeOfEndpointInstance typings are correct
null as TypeOfEndpointInstance<typeof endpointWithoutParam>['getStaticPath'];
// @dts-jest:pass:snap resulting TypeOfEndpointInstance typings are correct
null as RequiredKeys<TypeOfEndpointInstance<typeof endpointWithoutParam>>;
// @dts-jest:pass:snap resulting TypeOfEndpointInstance typings are correct
null as RequiredKeys<TypeOfEndpointInstance<typeof endpointWithoutParam>['Input']>;

// @dts-jest:pass:snap resulting TypeOfEndpointInstance typings are correct
null as TypeOfEndpointInstance<typeof endpointWithoutInput>['Errors'];
// @dts-jest:pass:snap resulting TypeOfEndpointInstance typings are correct
null as TypeOfEndpointInstance<typeof endpointWithoutInput>['Input'];
// @dts-jest:pass:snap resulting TypeOfEndpointInstance typings are correct
null as TypeOfEndpointInstance<typeof endpointWithoutInput>['Output'];
// @dts-jest:pass:snap resulting TypeOfEndpointInstance typings are correct
null as TypeOfEndpointInstance<typeof endpointWithoutInput>['getPath'];
// @dts-jest:pass:snap resulting TypeOfEndpointInstance typings are correct
null as TypeOfEndpointInstance<typeof endpointWithoutInput>['getStaticPath'];
// @dts-jest:pass:snap resulting TypeOfEndpointInstance typings are correct
null as RequiredKeys<TypeOfEndpointInstance<typeof endpointWithoutInput>>;
// @dts-jest:pass:snap resulting TypeOfEndpointInstance typings are correct
null as RequiredKeys<TypeOfEndpointInstance<typeof endpointWithoutInput>['Input']>;

// @dts-jest:pass:snap resulting TypeOfEndpointInstance typings are correct
null as TypeOfEndpointInstance<typeof endpointWithBody>['Errors'];
// @dts-jest:pass:snap resulting TypeOfEndpointInstance typings are correct
null as TypeOfEndpointInstance<typeof endpointWithBody>['Input'];
// @dts-jest:pass:snap resulting TypeOfEndpointInstance typings are correct
null as TypeOfEndpointInstance<typeof endpointWithBody>['Output'];
// @dts-jest:pass:snap resulting TypeOfEndpointInstance typings are correct
null as TypeOfEndpointInstance<typeof endpointWithBody>['getPath'];
// @dts-jest:pass:snap resulting TypeOfEndpointInstance typings are correct
null as TypeOfEndpointInstance<typeof endpointWithBody>['getStaticPath'];
// @dts-jest:pass:snap resulting TypeOfEndpointInstance typings are correct
null as RequiredKeys<TypeOfEndpointInstance<typeof endpointWithBody>>;
// @dts-jest:pass:snap resulting TypeOfEndpointInstance typings are correct
null as RequiredKeys<TypeOfEndpointInstance<typeof endpointWithBody>['Input']>;

// @dts-jest:pass:snap resulting TypeOfEndpointInstance typings are correct
null as TypeOfEndpointInstance<typeof endpointWithErrors>['Errors'];
// @dts-jest:pass:snap resulting TypeOfEndpointInstance typings are correct
null as TypeOfEndpointInstance<typeof endpointWithErrors>['Input'];
// @dts-jest:pass:snap resulting TypeOfEndpointInstance typings are correct
null as TypeOfEndpointInstance<typeof endpointWithErrors>['Output'];
// @dts-jest:pass:snap resulting TypeOfEndpointInstance typings are correct
null as TypeOfEndpointInstance<typeof endpointWithErrors>['getPath'];
// @dts-jest:pass:snap resulting TypeOfEndpointInstance typings are correct
null as TypeOfEndpointInstance<typeof endpointWithErrors>['getStaticPath'];
// @dts-jest:pass:snap resulting TypeOfEndpointInstance typings are correct
null as RequiredKeys<TypeOfEndpointInstance<typeof endpointWithErrors>>;
// @dts-jest:pass:snap resulting TypeOfEndpointInstance typings are correct
null as RequiredKeys<TypeOfEndpointInstance<typeof endpointWithErrors>['Input']>;
