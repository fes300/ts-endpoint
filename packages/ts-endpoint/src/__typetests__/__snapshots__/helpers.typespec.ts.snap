// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`resulting TypeOfEndpointInstance typings are correct (type) should match snapshot 1`] = `"{}"`;

exports[`resulting TypeOfEndpointInstance typings are correct (type) should match snapshot 2`] = `"{ Params: { id: number; }; Headers?: never; Query: { color: string; }; Body?: never; }"`;

exports[`resulting TypeOfEndpointInstance typings are correct (type) should match snapshot 3`] = `"{ crayons: string[]; }"`;

exports[`resulting TypeOfEndpointInstance typings are correct (type) should match snapshot 4`] = `"(args: { id: number; }) => string"`;

exports[`resulting TypeOfEndpointInstance typings are correct (type) should match snapshot 5`] = `"(f: (paramName: "id") => string) => string"`;

exports[`resulting TypeOfEndpointInstance typings are correct (type) should match snapshot 6`] = `""Input" | "Errors" | "getPath" | "Method" | "Output" | "getStaticPath""`;

exports[`resulting TypeOfEndpointInstance typings are correct (type) should match snapshot 7`] = `""Params" | "Query""`;

exports[`resulting TypeOfEndpointInstance typings are correct (type) should match snapshot 8`] = `"{}"`;

exports[`resulting TypeOfEndpointInstance typings are correct (type) should match snapshot 9`] = `"{ Params?: never; Headers?: never; Query: { color: string; }; Body?: never; }"`;

exports[`resulting TypeOfEndpointInstance typings are correct (type) should match snapshot 10`] = `"{ crayons: string[]; }"`;

exports[`resulting TypeOfEndpointInstance typings are correct (type) should match snapshot 11`] = `"(i?: {}) => string"`;

exports[`resulting TypeOfEndpointInstance typings are correct (type) should match snapshot 12`] = `"(i?: {}) => string"`;

exports[`resulting TypeOfEndpointInstance typings are correct (type) should match snapshot 13`] = `""Input" | "Errors" | "getPath" | "Method" | "Output" | "getStaticPath""`;

exports[`resulting TypeOfEndpointInstance typings are correct (type) should match snapshot 14`] = `""Query""`;

exports[`resulting TypeOfEndpointInstance typings are correct (type) should match snapshot 15`] = `"{}"`;

exports[`resulting TypeOfEndpointInstance typings are correct (type) should match snapshot 16`] = `"void"`;

exports[`resulting TypeOfEndpointInstance typings are correct (type) should match snapshot 17`] = `"{ crayons: string[]; }"`;

exports[`resulting TypeOfEndpointInstance typings are correct (type) should match snapshot 18`] = `"(i?: {}) => string"`;

exports[`resulting TypeOfEndpointInstance typings are correct (type) should match snapshot 19`] = `"(i?: {}) => string"`;

exports[`resulting TypeOfEndpointInstance typings are correct (type) should match snapshot 20`] = `""Input" | "Errors" | "getPath" | "Method" | "Output" | "getStaticPath""`;

exports[`resulting TypeOfEndpointInstance typings are correct (type) should match snapshot 21`] = `"never"`;

exports[`resulting TypeOfEndpointInstance typings are correct (type) should match snapshot 22`] = `"{}"`;

exports[`resulting TypeOfEndpointInstance typings are correct (type) should match snapshot 23`] = `"{ Params?: never; Headers?: never; Query?: never; Body: { id: number; }; }"`;

exports[`resulting TypeOfEndpointInstance typings are correct (type) should match snapshot 24`] = `"{ crayons: string[]; }"`;

exports[`resulting TypeOfEndpointInstance typings are correct (type) should match snapshot 25`] = `"(i?: {}) => string"`;

exports[`resulting TypeOfEndpointInstance typings are correct (type) should match snapshot 26`] = `"(i?: {}) => string"`;

exports[`resulting TypeOfEndpointInstance typings are correct (type) should match snapshot 27`] = `""Input" | "Errors" | "getPath" | "Method" | "Output" | "getStaticPath""`;

exports[`resulting TypeOfEndpointInstance typings are correct (type) should match snapshot 28`] = `""Body""`;

exports[`resulting TypeOfEndpointInstance typings are correct (type) should match snapshot 29`] = `"{ 401: undefined; 404: { message: string; }; 500: { foo: number; }; }"`;

exports[`resulting TypeOfEndpointInstance typings are correct (type) should match snapshot 30`] = `"{ Params: { id: number; }; Headers?: never; Query: { color: string; }; Body?: never; }"`;

exports[`resulting TypeOfEndpointInstance typings are correct (type) should match snapshot 31`] = `"{ crayons: string[]; }"`;

exports[`resulting TypeOfEndpointInstance typings are correct (type) should match snapshot 32`] = `"(args: { id: number; }) => string"`;

exports[`resulting TypeOfEndpointInstance typings are correct (type) should match snapshot 33`] = `"(f: (paramName: "id") => string) => string"`;

exports[`resulting TypeOfEndpointInstance typings are correct (type) should match snapshot 34`] = `""Input" | "Errors" | "getPath" | "Method" | "Output" | "getStaticPath""`;

exports[`resulting TypeOfEndpointInstance typings are correct (type) should match snapshot 35`] = `""Params" | "Query""`;
